pkg/kubelet/pod_workers.go

```diff
@@ -139,77 +364,233 @@ type podWorkers struct {
+	}
+
+	// once a pod is terminated by UID, it cannot reenter the pod worker (until the UID is purged by housekeeping)
+	if status.IsFinished() {
+		klog.V(4).InfoS("Pod is finished processing, no further updates", "pod", klog.KObj(pod), "podUID", pod.UID)
+		return
+	}
+
+	// check for a transition to terminating
+	var becameTerminating bool
+	if !status.IsTerminating() {
+		switch {
+		case isRuntimePod:
+			klog.V(4).InfoS("Pod is orphaned and must be torn down", "pod", klog.KObj(pod), "podUID", pod.UID)
+			status.deleted = true
+			status.terminatingAt = now
+			becameTerminating = true
+		case pod.DeletionTimestamp != nil:
+			klog.V(4).InfoS("Pod is marked for graceful deletion, begin teardown", "pod", klog.KObj(pod), "podUID", pod.UID)
+			status.deleted = true
+			status.terminatingAt = now
+			becameTerminating = true
+		case pod.Status.Phase == v1.PodFailed, pod.Status.Phase == v1.PodSucceeded:
+			klog.V(4).InfoS("Pod is in a terminal phase (success/failed), begin teardown", "pod", klog.KObj(pod), "podUID", pod.UID)
+			status.terminatingAt = now
+			becameTerminating = true
+		case options.UpdateType == kubetypes.SyncPodKill:
+			if options.KillPodOptions != nil && options.KillPodOptions.Evict {
+				klog.V(4).InfoS("Pod is being evicted by the kubelet, begin teardown", "pod", klog.KObj(pod), "podUID", pod.UID)
+				status.evicted = true
+			} else {
+				klog.V(4).InfoS("Pod is being removed by the kubelet, begin teardown", "pod", klog.KObj(pod), "podUID", pod.UID)
+			}
+			status.terminatingAt = now
+			becameTerminating = true
+		}
+	}
+
+	// once a pod is terminating, all updates are kills and the grace period can only decrease
+	var workType PodWorkType
+	var wasGracePeriodShortened bool
+	switch {
+	case status.IsTerminated():
+		workType = TerminatedPodWork
+
+		if options.KillPodOptions != nil {
+			if ch := options.KillPodOptions.CompletedCh; ch != nil {
+				close(ch)
+			}
+		}
+		options.KillPodOptions = nil
+
+	case status.IsTerminating():
+		workType = TerminatingPodWork
+		if options.KillPodOptions == nil {
+			options.KillPodOptions = &KillPodOptions{}
+		}
+
+		if ch := options.KillPodOptions.CompletedCh; ch != nil {
+			status.notifyPostTerminating = append(status.notifyPostTerminating, ch)
+		}
+
+		gracePeriod, gracePeriodShortened := calculateEffectiveGracePeriod(status, pod, options.KillPodOptions)
+
+		wasGracePeriodShortened = gracePeriodShortened
+		status.gracePeriod = gracePeriod
+		// always set the grace period for syncTerminatingPod so we don't have to recalculate,
+		// will never be zero.
+		options.KillPodOptions.PodTerminationGracePeriodSecondsOverride = &gracePeriod
+
+		// if a static pod comes through, start tracking it explicitly (cleared by the pod worker loop)
+		if kubelettypes.IsStaticPod(pod) {
+			p.terminatingStaticPodFullnames[kubecontainer.GetPodFullName(pod)] = struct{}{}
+		}
+
+	default:
+		workType = SyncPodWork
+
+		// completely ignore KillPodOptions
```

Comment:
When could this happen? Should we return an error instead?

Reply:
We could simply have another public method besides UpdatePod which is "kill pod"

Reply:
Let's at least add comment to point out that this is not a valid case.